indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,256 @@
+      return -1;
+    }
+
+    final int[] validTaskCounts = CostBasedAutoScaler.computeValidTaskCounts(partitionCount, currentTaskCount);
+
+    if (validTaskCounts.length == 0) {
+      log.warn("No valid task counts after applying constraints for supervisorId [%s]", supervisorId);
+      return -1;
+    }
+
+    // If idle is already in the ideal range [0.2, 0.6], optimal utilization has been achieved.
+    // No scaling is needed - maintain stability by staying at current task count.
+    final double currentIdleRatio = metrics.getPollIdleRatio();
+    if (currentIdleRatio >= 0 && WeightedCostFunction.isIdleInIdealRange(currentIdleRatio)) {
+      log.info(
+          "Idle ratio [%.3f] is in ideal range for supervisorId [%s], no scaling needed",
+          currentIdleRatio,
+          supervisorId
+      );
+      return -1;
+    }
+
+    int optimalTaskCount = -1;
+    double optimalCost = Double.POSITIVE_INFINITY;
+
+    for (int taskCount : validTaskCounts) {
+      double cost = costFunction.computeCost(metrics, taskCount, config);
+      log.debug("Proposed task count: %d, Cost: %.4f", taskCount, cost);
+      if (cost < optimalCost) {
+        optimalTaskCount = taskCount;
+        optimalCost = cost;
+      }
+    }
+
+    emitter.emit(metricBuilder.setMetric(OPTIMAL_TASK_COUNT_METRIC, (long) optimalTaskCount));
+
+    log.debug(
+        "Cost-based scaling evaluation for supervisorId [%s]: current=%d, optimal=%d, cost=%.4f, "
+        + "avgPartitionLag=%.2f, pollIdleRatio=%.3f",
+        supervisorId,
+        metrics.getCurrentTaskCount(),
+        optimalTaskCount,
+        optimalCost,
+        metrics.getAvgPartitionLag(),
+        metrics.getPollIdleRatio()
+    );
+
+    if (optimalTaskCount == currentTaskCount) {
+      return -1;
+    }
+    // Perform both scale-up and scale-down proactively
+    // Future versions may perform scale-down on task rollover only
+    return optimalTaskCount;
+  }
+
+  /**
+   * Generates valid task counts based on partitions-per-task ratios.
+   * This enables gradual scaling and avoids large jumps.
+   * Limits the range of task counts considered to avoid excessive computation.
+   *
+   * @return sorted list of valid task counts within bounds
+   */
+  static int[] computeValidTaskCounts(int partitionCount, int currentTaskCount)
+  {
+    if (partitionCount <= 0) {
+      return new int[]{};
+    }
+
+    List<Integer> result = new ArrayList<>();
+    final int currentPartitionsPerTask = partitionCount / currentTaskCount;
+    // Minimum partitions per task corresponds to maximum number of tasks (scale up) and vice versa.
+    final int minPartitionsPerTask = Math.max(1, currentPartitionsPerTask - MAX_INCREASE_IN_PARTITIONS_PER_TASK);
+    final int maxPartitionsPerTask = Math.min(
+        partitionCount,
+        Math.max(
+            minPartitionsPerTask,
+            currentPartitionsPerTask + MIN_INCREASE_IN_PARTITIONS_PER_TASK
+        )
+    );
```

Comment:
Nit: This may be simplified since the limits are already handled properly by the `for` loop.

```suggestion
    final int maxPartitionsPerTask = Math.min(partitionCount, currentPartitionsPerTask + MAX_INCREASE_IN_PARTITIONS_PER_TASK);
```